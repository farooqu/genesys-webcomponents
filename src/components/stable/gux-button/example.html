<h1>gux-button</h1>

<h2>Enabled</h2>
<div class="container">
  <gux-button onclick="notify(event)">Default</gux-button>
  <gux-button accent="primary" onclick="notify(event)">Primary</gux-button>
  <gux-button accent="secondary" onclick="notify(event)">Secondary</gux-button>
  <gux-button accent="tertiary" onclick="notify(event)">Tertiary</gux-button>
  <gux-button accent="invalid" onclick="notify(event)"
    >Invalid accent</gux-button
  >
  <gux-button onclick="notify(event)">
    <span>Text in a span element</span>
  </gux-button>
  <gux-button onclick="notify(event)">
    <gux-icon
      icon-name="trashcan"
      screenreader-text="This will be read by a screen reader"
    ></gux-icon>
  </gux-button>
  <gux-button
    gux-title="This will be read by a screen reader"
    onclick="notify(event)"
  >
    <gux-icon icon-name="trashcan" decorative></gux-icon>
  </gux-button>
  <gux-button onclick="notify(event)">
    <gux-icon icon-name="trashcan" decorative></gux-icon>
    <span>Text in a span element</span>
  </gux-button>
  <gux-button onclick="notify(event)">
    <span>Text in a span element</span>
    <span>More Text in a span element</span>
  </gux-button>
  <gux-button onclick="notify(event)">
    <span>
      <span>Text in a span element</span>
      <span>More Text in a span element</span>
    </span>
  </gux-button>
  <gux-button style="width: 100px" onclick="notify(event)">
    <span>Text in a span element</span>
    <span>More Text in a span element</span>
  </gux-button>

  <gux-button style="width: 100px" onclick="notify(event)">
    Text in a span element
  </gux-button>
</div>

<h2>Disabled</h2>
<div class="container">
  <gux-button disabled onclick="notify(event)">Default</gux-button>
  <gux-button accent="primary" disabled onclick="notify(event)"
    >Primary</gux-button
  >
  <gux-button accent="secondary" disabled onclick="notify(event)"
    >Secondary</gux-button
  >
  <gux-button accent="tertiary" disabled onclick="notify(event)"
    >Tertiary</gux-button
  >
  <gux-button accent="invalid" disabled onclick="notify(event)"
    >Invalid accent</gux-button
  >
  <gux-button disabled onclick="notify(event)">
    <span>Text in a span element</span>
  </gux-button>
  <gux-button disabled onclick="notify(event)">
    <gux-icon
      icon-name="trashcan"
      screenreader-text="This will be read by a screen reader"
    ></gux-icon>
  </gux-button>
  <gux-button
    disabled
    gux-title="This will be read by a screen reader"
    onclick="notify(event)"
  >
    <gux-icon icon-name="trashcan" decorative></gux-icon>
  </gux-button>
  <gux-button disabled onclick="notify(event)">
    <gux-icon icon-name="trashcan" decorative></gux-icon>
    <span>Text in a span element</span>
  </gux-button>
  <gux-button disabled onclick="notify(event)">
    <span>Text in a span element</span>
    <span>More Text in a span element</span>
  </gux-button>
  <gux-button disabled onclick="notify(event)">
    <span>
      <span>Text in a span element</span>
      <span>More Text in a span element</span>
    </span>
  </gux-button>
  <gux-button style="width: 100px" disabled onclick="notify(event)">
    <span>Text in a span element</span>
    <span>More Text in a span element</span>
  </gux-button>

  <gux-button style="width: 100px" disabled onclick="notify(event)">
    Text in a span element
  </gux-button>
</div>

<style>
  .container {
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    align-content: stretch;
    justify-content: flex-start;
  }

  gux-button {
    margin: 5px;
  }
</style>
