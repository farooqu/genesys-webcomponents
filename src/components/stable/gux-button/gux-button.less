@import (reference) '../../../style/color.less';
@import (reference) '../../../style/mixins.less';
@import (reference) '../../../style/typography.less';
@import (reference) '../../../style/spacing.less';

:host {
  display: inline-block;
  min-width: 32px;
  pointer-events: none;
}

:host(:focus) {
  outline: none;

  button {
    outline: none;
  }
}

::slotted(gux-icon) {
  width: 16px;
  height: 16px;
}

::slotted(*) {
  padding-left: @spacing-2xs;
  vertical-align: middle;
}

::slotted(*:first-child) {
  padding: 0;
}

button {
  width: 100%;
  height: 32px;
  padding: 0 @spacing-medium;
  overflow: hidden;
  color: @gux-black-50;
  text-overflow: ellipsis;
  white-space: nowrap;
  pointer-events: auto;
  cursor: pointer;
  background-color: @gux-grey-60;
  border: none;
  border-radius: 4px;
  .body-font-bold();

  &[disabled] {
    color: fade(@gux-black-50, 50%);
    cursor: default;
    background-color: fade(@gux-grey-60, 50%);
  }

  &:hover:enabled {
    background-color: @gux-grey-50;
  }

  &:active:enabled {
    background-color: @gux-grey-40;
  }

  &:focus:enabled {
    .focus-ring();
  }

  &.gux-tertiary {
    color: @gux-blue-60;
    background-color: @gux-grey-100;
    border: 1px solid @gux-blue-60;

    &[disabled] {
      color: fade(@gux-blue-60, 50%);
      background-color: fade(@gux-grey-100, 50%);
      border-color: fade(@gux-blue-60, 50%);
    }

    &:hover:enabled {
      color: @gux-grey-100;
      background-color: @gux-blue-50;
    }

    &:active:enabled {
      color: @gux-grey-100;
      background-color: @gux-blue-40;
    }
  }

  &.gux-primary {
    color: @gux-grey-100;
    background-color: @gux-blue-60;

    &[disabled] {
      color: fade(@gux-grey-100, 50%);
      background-color: fade(@gux-blue-60, 50%);
    }

    &:hover:enabled {
      background-color: @gux-blue-50;
    }

    &:active:enabled {
      background-color: @gux-blue-40;
    }
  }
}
