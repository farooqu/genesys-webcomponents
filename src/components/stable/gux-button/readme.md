# gux-button

This custom component is a simple button having some styling on it.
You can choose between two type (secondary and primary).

<!-- Auto Generated Below -->


## Properties

| Property   | Attribute   | Description                                  | Type                                     | Default       |
| ---------- | ----------- | -------------------------------------------- | ---------------------------------------- | ------------- |
| `accent`   | `accent`    | The component accent (secondary or primary). | `"primary" \| "secondary" \| "tertiary"` | `'secondary'` |
| `disabled` | `disabled`  | Indicate if the button is disabled or not    | `boolean`                                | `false`       |
| `guxTitle` | `gux-title` | The component title                          | `string`                                 | `undefined`   |
| `type`     | `type`      | The component button type                    | `"button" \| "reset" \| "submit"`        | `'button'`    |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
