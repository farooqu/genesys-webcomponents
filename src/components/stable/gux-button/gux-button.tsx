import { Component, Element, h, JSX, Prop } from '@stencil/core';

import { trackComponent } from '../../../usage-tracking';
import { GuxButtonAccent, GuxButtonType } from './gux-button.types';

@Component({
  styleUrl: 'gux-button.less',
  tag: 'gux-button',
  shadow: { delegatesFocus: true }
})
export class GuxButton {
  @Element()
  private root: HTMLElement;

  /**
   * The component button type
   */
  @Prop()
  type: GuxButtonType = 'button';

  /**
   * The component title
   */
  @Prop()
  guxTitle: string;

  /**
   * Indicate if the button is disabled or not
   */
  @Prop()
  disabled: boolean = false;

  /**
   * The component accent (secondary or primary).
   */
  @Prop()
  accent: GuxButtonAccent = 'secondary';

  componentWillLoad() {
    trackComponent(this.root, { variant: this.accent });
    this.makeSlotContentDisableable();
  }
  render(): JSX.Element {
    return (
      <button
        type={this.type}
        title={this.guxTitle}
        disabled={this.disabled}
        class={`gux-${this.accent}`}
      >
        <slot />
      </button>
    ) as JSX.Element;
  }

  private makeSlotContentDisableable() {
    this.root.shadowRoot.addEventListener(
      'click',
      (event: MouseEvent): void => {
        if (this.disabled) {
          event.stopImmediatePropagation();
          event.stopPropagation();
          event.preventDefault();
        }
      }
    );
    Array.from(this.root.children).forEach(slotElement => {
      slotElement.addEventListener('click', (event: MouseEvent): void => {
        if (this.disabled) {
          event.stopImmediatePropagation();
          event.stopPropagation();
          event.preventDefault();
        }
      });
    });
  }
}
