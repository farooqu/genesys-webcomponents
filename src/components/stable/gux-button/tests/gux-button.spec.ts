import { newSpecPage } from '@stencil/core/testing';
import { GuxButton } from '../../gux-button/gux-button';

const components = [GuxButton];
const language = 'en';

describe('gux-button', () => {
  describe('#render', () => {
    [
      {
        description: 'should render default button',
        html: '<gux-button>Button</gux-button>'
      },
      {
        description: 'should render primary button',
        html: '<gux-button accent="primary">Button</gux-button>'
      },
      {
        description: 'should render secondary button',
        html: '<gux-button accent="secondary">Button</gux-button>'
      },
      {
        description: 'should render tertiary button',
        html: '<gux-button accent="tertiary">Button</gux-button>'
      },
      {
        description: 'should render invalid accent button',
        html: '<gux-button accent="invalid">Invalid accent</gux-button>'
      },
      {
        description: 'should render disabled default button',
        html: '<gux-button disabled>Button</gux-button>'
      },
      {
        description: 'should render disabled primary button',
        html: '<gux-button accent="primary" disabled>Button</gux-button>'
      },
      {
        description: 'should render disabled secondary button',
        html: '<gux-button accent="secondary" disabled>Button</gux-button>'
      },
      {
        description: 'should render disabled tertiary button',
        html: '<gux-button accent="tertiary" disabled>Button</gux-button>'
      },
      {
        description: 'should render disabled invalid accent button',
        html: '<gux-button accent="invalid" disabled>Invalid accent</gux-button>'
      }
    ].forEach(({ description, html }) => {
      it(description, async () => {
        const page = await newSpecPage({ components, html, language });

        expect(page.rootInstance).toBeInstanceOf(GuxButton);
        expect(page.root).toMatchSnapshot();
      });
    });
  });

  describe('click', () => {
    it('should fire a click event when an enabled button is clicked', async () => {
      const html = '<gux-button>Button</gux-button>';
      const page = await newSpecPage({ components, html, language });
      const element = page.root as HTMLElement;
      const clickSpy = jest.fn();

      element.addEventListener('click', clickSpy);

      element.click();
      await page.waitForChanges();

      expect(clickSpy).toHaveBeenCalled();
    });

    it('should fire a click event when an enabled button slot content is clicked', async () => {
      const html = '<gux-button><span>Span</span></gux-button>';
      const page = await newSpecPage({ components, html, language });
      const element = page.root as HTMLElement;
      const span = page.root.querySelector('span') as HTMLButtonElement;
      const clickSpy = jest.fn();

      element.addEventListener('click', clickSpy);

      span.click();
      await page.waitForChanges();

      expect(clickSpy).toHaveBeenCalled();
    });

    it('should not fire a click event when a disabled button slot content is clicked', async () => {
      const html = '<gux-button disabled><span>Span</span></gux-button>';
      const page = await newSpecPage({ components, html, language });
      const element = page.root as HTMLElement;
      const span = page.root.querySelector('span') as HTMLElement;
      const clickSpy = jest.fn();

      element.addEventListener('click', clickSpy);

      span.click();
      await page.waitForChanges();

      expect(clickSpy).not.toHaveBeenCalled();
    });
  });
});
