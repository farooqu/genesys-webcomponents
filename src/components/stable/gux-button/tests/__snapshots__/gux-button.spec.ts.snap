// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gux-button #render should render default button 1`] = `
<gux-button>
  <mock:shadow-root>
    <button class="gux-secondary" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
  Button
</gux-button>
`;

exports[`gux-button #render should render disabled default button 1`] = `
<gux-button disabled="">
  <mock:shadow-root>
    <button class="gux-secondary" disabled="" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
  Button
</gux-button>
`;

exports[`gux-button #render should render disabled invalid accent button 1`] = `
<gux-button accent="invalid" disabled="">
  <mock:shadow-root>
    <button class="gux-invalid" disabled="" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
  Invalid accent
</gux-button>
`;

exports[`gux-button #render should render disabled primary button 1`] = `
<gux-button accent="primary" disabled="">
  <mock:shadow-root>
    <button class="gux-primary" disabled="" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
  Button
</gux-button>
`;

exports[`gux-button #render should render disabled secondary button 1`] = `
<gux-button accent="secondary" disabled="">
  <mock:shadow-root>
    <button class="gux-secondary" disabled="" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
  Button
</gux-button>
`;

exports[`gux-button #render should render disabled tertiary button 1`] = `
<gux-button accent="tertiary" disabled="">
  <mock:shadow-root>
    <button class="gux-tertiary" disabled="" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
  Button
</gux-button>
`;

exports[`gux-button #render should render invalid accent button 1`] = `
<gux-button accent="invalid">
  <mock:shadow-root>
    <button class="gux-invalid" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
  Invalid accent
</gux-button>
`;

exports[`gux-button #render should render primary button 1`] = `
<gux-button accent="primary">
  <mock:shadow-root>
    <button class="gux-primary" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
  Button
</gux-button>
`;

exports[`gux-button #render should render secondary button 1`] = `
<gux-button accent="secondary">
  <mock:shadow-root>
    <button class="gux-secondary" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
  Button
</gux-button>
`;

exports[`gux-button #render should render tertiary button 1`] = `
<gux-button accent="tertiary">
  <mock:shadow-root>
    <button class="gux-tertiary" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
  Button
</gux-button>
`;
